### PR TITLE
Allow for setting client ID / secret through files

### PR DESCRIPTION
**Pull request by mpminardi**
Adds the ability to specify a path to the client ID and client secret in place of supplying their values directly in the provider.

### CODE REVIEW

File: internal/provider/provider_test.go [hunk 1]
```diff
@@ -223,11 +223,11 @@ func TestAccProvider_clientSecretAuth(t *testing.T) {
...
 	}
-	if os.Getenv("ARM_CLIENT_ID") == "" {
-		t.Skip("ARM_CLIENT_ID not set")
+	if os.Getenv("ARM_CLIENT_ID") == "" && os.Getenv("ARM_CLIENT_ID_PATH") == "" {
+		t.Skip("ARM_CLIENT_ID or ARM_CLIENT_ID_PATH not set")
 	}
-	if os.Getenv("ARM_CLIENT_SECRET") == "" {
-		t.Skip("ARM_CLIENT_SECRET not set")
+	if os.Getenv("ARM_CLIENT_SECRET") == "" && os.Getenv("ARM_CLIENT_SECRET_PATH") == "" {
+		t.Skip("ARM_CLIENT_SECRET or ARM_CLIENT_SECRET_PATH not set")
```
Comment (tombuildsstuff) [inline internal/provider/provider_test.go:230]: could we split this out as a separate test?

Comment (mpminardi) [reply internal/provider/provider_test.go:230]: For sure! I'm assuming you mean having one test check `ARM_CLIENT_ID` with `ARM_CLIENT_SECRET` and another to do `ARM_CLIENT_ID_PATH` with `ARM_CLIENT_SECRET_PATH`?

Comment (tombuildsstuff) [reply internal/provider/provider_test.go]: yeah exactly 👍 - thanks

File: internal/provider/provider.go [hunk 2]
```diff
@@ -178,6 +178,13 @@ func azureProvider(supportLegacyTestSuite bool) *schema.Provider {
 				Description: "The Client ID which should be used.",
 			},
 
+			"client_id_path": {
```
Comment (manicminer) [inline internal/provider/provider.go:181]:
```suggestion
			"client_id_file_path": {
```

File: internal/provider/provider.go [hunk 3]
```diff
@@ -178,6 +178,13 @@ func azureProvider(supportLegacyTestSuite bool) *schema.Provider {
...
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("ARM_CLIENT_ID_PATH", ""),
+				Description: "Path to a file containing the Client ID which should be used.",
```
Comment (manicminer) [inline internal/provider/provider.go:185]:
```suggestion
				Description: "The path to a file containing the Client ID which should be used.",
```

File: internal/provider/provider.go [hunk 4]
```diff
@@ -178,6 +178,13 @@ func azureProvider(supportLegacyTestSuite bool) *schema.Provider {
 				Description: "The Client ID which should be used.",
 			},
 
+			"client_id_path": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("ARM_CLIENT_ID_PATH", ""),
```
Comment (manicminer) [inline internal/provider/provider.go:184]:
```suggestion
				DefaultFunc: schema.EnvDefaultFunc("ARM_CLIENT_ID_FILE_PATH", ""),
```

File: internal/provider/provider.go [hunk 5]
```diff
@@ -238,6 +245,13 @@ func azureProvider(supportLegacyTestSuite bool) *schema.Provider {
 				Description: "The Client Secret which should be used. For use When authenticating as a Service Principal using a Client Secret.",
 			},
 
+			"client_secret_path": {
```
Comment (manicminer) [inline internal/provider/provider.go:248]:
```suggestion
			"client_secret_file_path": {
```

File: internal/provider/provider.go [hunk 6]
```diff
@@ -238,6 +245,13 @@ func azureProvider(supportLegacyTestSuite bool) *schema.Provider {
...
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("ARM_CLIENT_SECRET_PATH", ""),
```
Comment (manicminer) [inline internal/provider/provider.go:251]:
```suggestion
				DefaultFunc: schema.EnvDefaultFunc("ARM_CLIENT_SECRET_FILE_PATH", ""),
```

File: internal/provider/provider.go [hunk 7]
```diff
@@ -373,6 +387,16 @@ func providerConfigure(p *schema.Provider) schema.ConfigureContextFunc {
 			return nil, diag.FromErr(err)
 		}
 
+		clientSecret, err := getClientSecret(d)
+		if err != nil {
+			return nil, diag.FromErr(err)
+		}
+
+		clientID, err := getClientID(d)
```
Comment (manicminer) [inline internal/provider/provider.go:395]:
```suggestion
		clientId, err := getClientId(d)
```

File: internal/provider/provider.go [hunk 8]
```diff
@@ -396,14 +420,14 @@ func providerConfigure(p *schema.Provider) schema.ConfigureContextFunc {
 
 		authConfig := &auth.Credentials{
 			Environment:        *env,
-			ClientID:           d.Get("client_id").(string),
+			ClientID:           *clientID,
```
Comment (manicminer) [inline internal/provider/provider.go:423]:
```suggestion
			ClientID:           *clientId,
```

File: internal/provider/provider.go [hunk 9]
```diff
@@ -505,6 +529,50 @@ func getOidcToken(d *schema.ResourceData) (*string, error) {
 	return &idToken, nil
 }
 
+func getClientID(d *schema.ResourceData) (*string, error) {
```
Comment (manicminer) [inline internal/provider/provider.go:532]:
```suggestion
func getClientId(d *schema.ResourceData) (*string, error) {
```

File: internal/provider/provider.go [hunk 10]
```diff
@@ -505,6 +529,50 @@ func getOidcToken(d *schema.ResourceData) (*string, error) {
...
+	clientID := strings.TrimSpace(d.Get("client_id").(string))
```
Comment (manicminer) [inline internal/provider/provider.go:533]:
```suggestion
	clientId := strings.TrimSpace(d.Get("client_id").(string))
```

File: internal/provider/provider.go [hunk 11]
```diff
@@ -505,6 +529,50 @@ func getOidcToken(d *schema.ResourceData) (*string, error) {
...
+
+	if path := d.Get("client_id_path").(string); path != "" {
```
Comment (manicminer) [inline internal/provider/provider.go:535]:
```suggestion
	if path := d.Get("client_id_file_path").(string); path != "" {
```

File: internal/provider/provider.go [hunk 12]
```diff
@@ -505,6 +529,50 @@ func getOidcToken(d *schema.ResourceData) (*string, error) {
...
+		fileClientIDRaw, err := os.ReadFile(path)
```
Comment (manicminer) [inline internal/provider/provider.go:536]:
```suggestion
		fileClientIdRaw, err := os.ReadFile(path)
```

File: internal/provider/provider.go [hunk 13]
```diff
@@ -505,6 +529,50 @@ func getOidcToken(d *schema.ResourceData) (*string, error) {
...
+
+		if err != nil {
+			return nil, fmt.Errorf("reading Client Secret from file %q: %v", path, err)
```
Comment (manicminer) [inline internal/provider/provider.go:539]:
```suggestion
			return nil, fmt.Errorf("reading Client ID from file %q: %v", path, err)
```

File: internal/provider/provider.go [hunk 14]
```diff
@@ -505,6 +529,50 @@ func getOidcToken(d *schema.ResourceData) (*string, error) {
...
+		}
+
+		fileClientID := strings.TrimSpace(string(fileClientIDRaw))
+
+		if clientID != "" && clientID != fileClientID {
```
Comment (manicminer) [inline internal/provider/provider.go:544]:
```suggestion
		if clientId != "" && clientId != fileClientId {
```

File: internal/provider/provider.go [hunk 15]
```diff
@@ -505,6 +529,50 @@ func getOidcToken(d *schema.ResourceData) (*string, error) {
...
+	clientID := strings.TrimSpace(d.Get("client_id").(string))
+
+	if path := d.Get("client_id_path").(string); path != "" {
+		fileClientIDRaw, err := os.ReadFile(path)
+
+		if err != nil {
+			return nil, fmt.Errorf("reading Client Secret from file %q: %v", path, err)
+		}
+
+		fileClientID := strings.TrimSpace(string(fileClientIDRaw))
```
Comment (manicminer) [inline internal/provider/provider.go:542]:
```suggestion
		fileClientId := strings.TrimSpace(string(fileClientIdRaw))
```

File: internal/provider/provider.go [hunk 16]
```diff
@@ -505,6 +529,50 @@ func getOidcToken(d *schema.ResourceData) (*string, error) {
...
+
+		if clientID != "" && clientID != fileClientID {
+			return nil, fmt.Errorf("mismatch between supplied Client ID and supplied Client ID file contents - please either remove one or ensure they match")
+		}
+
+		clientID = fileClientID
```
Comment (manicminer) [inline internal/provider/provider.go:548]:
```suggestion
		clientId = fileClientId
```

File: internal/provider/provider.go [hunk 17]
```diff
@@ -505,6 +529,50 @@ func getOidcToken(d *schema.ResourceData) (*string, error) {
...
+	}
+
+	return &clientID, nil
```
Comment (manicminer) [inline internal/provider/provider.go:551]:
```suggestion
	return &clientId, nil
```

File: internal/provider/provider.go [hunk 18]
```diff
@@ -505,6 +529,50 @@ func getOidcToken(d *schema.ResourceData) (*string, error) {
...
+}
+
+func getClientSecret(d *schema.ResourceData) (*string, error) {
+	clientSecret := strings.TrimSpace(d.Get("client_secret").(string))
+
+	if path := d.Get("client_secret_path").(string); path != "" {
```
Comment (manicminer) [inline internal/provider/provider.go:557]:
```suggestion
	if path := d.Get("client_secret_file_path").(string); path != "" {
```

File: internal/provider/provider_test.go [hunk 19]
```diff
@@ -230,6 +230,11 @@ func TestAccProvider_clientSecretAuth(t *testing.T) {
 		t.Skip("ARM_CLIENT_SECRET not set")
 	}
 
+	// Ensure we are running using the expected env-vars
+	// t.SetEnv does automatic cleanup / resets the values after the test
+	t.Setenv("ARM_CLIENT_ID_PATH", "")
+	t.Setenv("ARM_CLIENT_SECRET_PATH", "")
```
Comment (manicminer) [inline internal/provider/provider_test.go:236]:
```suggestion
	t.Setenv("ARM_CLIENT_ID_FILE_PATH", "")
	t.Setenv("ARM_CLIENT_SECRET_FILE_PATH", "")
```

File: internal/provider/provider_test.go [hunk 20]
```diff
@@ -244,12 +249,85 @@ func TestAccProvider_clientSecretAuth(t *testing.T) {
 			t.Fatalf("configuring environment %q: %v", envName, err)
 		}
 
+		clientID, err := getClientID(d)
```
Comment (manicminer) [inline internal/provider/provider_test.go:252]:
```suggestion
		clientId, err := getClientId(d)
```

File: internal/provider/provider_test.go [hunk 21]
```diff
@@ -244,12 +249,85 @@ func TestAccProvider_clientSecretAuth(t *testing.T) {
...
+
+		clientSecret, err := getClientSecret(d)
+		if err != nil {
+			return nil, diag.FromErr(err)
+		}
+
+		authConfig := &auth.Credentials{
+			Environment:                           *env,
+			TenantID:                              d.Get("tenant_id").(string),
+			ClientID:                              *clientID,
```
Comment (manicminer) [inline internal/provider/provider_test.go:265]:
```suggestion
			ClientID:                              *clientId,
```

File: internal/provider/provider_test.go [hunk 22]
```diff
@@ -244,12 +249,85 @@ func TestAccProvider_clientSecretAuth(t *testing.T) {
...
+	}
+
+	if errs := testCheckProvider(provider); len(errs) > 0 {
+		for _, err := range errs {
+			t.Error(err)
+		}
+	}
+}
+
+func TestAccProvider_clientSecretAuth_fileBased(t *testing.T) {
```
Comment (manicminer) [inline internal/provider/provider_test.go:285]:
```suggestion
func TestAccProvider_clientSecretAuthFromFiles(t *testing.T) {
```

File: internal/provider/provider_test.go [hunk 23]
```diff
@@ -244,12 +249,85 @@ func TestAccProvider_clientSecretAuth(t *testing.T) {
...
+	if os.Getenv("TF_ACC") == "" {
+		t.Skip("TF_ACC not set")
+	}
+	if os.Getenv("ARM_CLIENT_ID_PATH") == "" {
+		t.Skip("ARM_CLIENT_ID_PATH not set")
+	}
+	if os.Getenv("ARM_CLIENT_SECRET_PATH") == "" {
+		t.Skip("ARM_CLIENT_SECRET_PATH not set")
+	}
```
Comment (manicminer) [inline internal/provider/provider_test.go:294]:
```suggestion
	if os.Getenv("ARM_CLIENT_ID_FILE_PATH") == "" {
		t.Skip("ARM_CLIENT_ID_FILE_PATH not set")
	}
	if os.Getenv("ARM_CLIENT_SECRET_FILE_PATH") == "" {
		t.Skip("ARM_CLIENT_SECRET_FILE_PATH not set")
	}
```

File: internal/provider/provider_test.go [hunk 24]
```diff
@@ -244,12 +249,85 @@ func TestAccProvider_clientSecretAuth(t *testing.T) {
...
+
+	// Support only Client Certificate authentication
+	provider.ConfigureContextFunc = func(ctx context.Context, d *schema.ResourceData) (interface{}, diag.Diagnostics) {
+		envName := d.Get("environment").(string)
+		env, err := environments.FromName(envName)
+		if err != nil {
+			t.Fatalf("configuring environment %q: %v", envName, err)
+		}
+
+		clientID, err := getClientID(d)
```
Comment (manicminer) [inline internal/provider/provider_test.go:315]:
```suggestion
		clientId, err := getClientId(d)
```

File: internal/provider/provider_test.go [hunk 25]
```diff
@@ -244,12 +249,85 @@ func TestAccProvider_clientSecretAuth(t *testing.T) {
...
+		clientSecret, err := getClientSecret(d)
+		if err != nil {
+			return nil, diag.FromErr(err)
+		}
+
 		authConfig := &auth.Credentials{
 			Environment:                           *env,
 			TenantID:                              d.Get("tenant_id").(string),
-			ClientID:                              d.Get("client_id").(string),
+			ClientID:                              *clientID,
```
Comment (manicminer) [inline internal/provider/provider_test.go:328]:
```suggestion
			ClientID:                              *clientId,
```

**Review by manicminer — CHANGES_REQUESTED** [review]
Thanks @mpminardi, happy to add this functionality. This change mostly looks good, I've tweaked & renamed a few things for consistency. I think we're going to have to run the tests using `t.SetEnv()` serially as env vars are process-wide, I'll add this similarly to how we do the same with other acceptance tests.

**Comment by mpminardi** [comment]
> Thanks @mpminardi, happy to add this functionality. This change mostly looks good, I've tweaked & renamed a few things for consistency. I think we're going to have to run the tests using `t.SetEnv()` serially as env vars are process-wide, I'll add this similarly to how we do the same with other acceptance tests.

Sounds good, thank you for the cleanup and fixing the tests!

**Review by jackofallops — APPROVED** [review]
Thanks @mpminardi we and @manicminer - This LGTM 👍

**Pull request merged**